Exclude OPTIONS for CORS preflight and Bearer Authorization

pedroneil opened the pull request:
Please consider this pull request that provides a solution to #627 

Adds 2 servlet init parameters to complement excludePatterns
excludeBearerAuthorization
excludeCorsPreflight

Omit OPTIONS requests but be more specific as they must be CORS preflight OPTIONS and not any old OPTIONS requests where Container Authorization allows for method ommission using the element - [http-method-omission](https://docs.oracle.com/javaee/6/tutorial/doc/gmmku.html) and this can let other unwanted requests through. 

A request is a CORS preflight request if the method is OPTIONS, and constains the HEADERS
Method: OPTIONS
Headers:
Origin
Access-Control-Request-Method
Access-Control-Request-Headers

If sending Basic Authorization accross domains, then there will be a preflight request with Access-Control-Request-Method: credentials which means that we want the filter to allow the request to get to CORS aware service

It may be desirable to only progress to the excludePatterns if there isn't any Bearer Authorization, this is especially the case if there is a mix of Windows, Bearer and Basic services on the Endpoint Services

hazendaz commented on Source/JNA/waffle-jna/pom.xml:
```diff
@@ -80,5 +80,11 @@
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>RELEASE</version>
```
@pedroneil Can you add proper version of mockito in property section?  The RELEASE and others like it are deprecated and considered bad practice.  Thanks.

pedroneil replied:
busy doing that

hazendaz commented on build.jdk8u172.cmd:
```diff
@@ -0,0 +1,12 @@
+@echo off
```
@pedroneil This probably wouldn't work for everyone.  You might want to pull this back for now.

pedroneil replied:
I'll remove build.jdk8u172.cmd

hazendaz commented:
@pedroneil Sorry I had missed this.  Other than a few formatting concerns I can address later and the couple things I did note, is this otherwise good to go?  I am looking to get a 1.9.1 release out shortly in order to get part of the other PR outstanding.  If this is good, I'd like to incorporate that as well.

pedroneil commented:
@hazendaz yes it is good and working in live testing

---

It was being used in testing before I created this PR; Waffle version 1.9.1-SNAPSHOT, built with JDK 1.8.0u172.

I would really benefit from being able to use a non Snapshot version of 1.9.1

I have added an entry to the FAQ.md and added [OAUTHT2 CORS FAQ page](faq/oauth2CORSBasicAndSSO.md)

1. after version 1.9.1 I propose some additional Guarding for the excludePatterns, see below
2. I will document the CorsAwareNegotiateSecurityFilter.java at a later date, this servlet was used while I was writting the enhancement to NegotiateSecurityFilter.

--- 

While writting the FAQ I've realised that there could be a expansion to the the excludePatterns specifying required Authorization Schemes to a baclward compatible Authorization Scheme prefix
e.g. 
```xml
<init-param>
  <param-name>excludePatterns
  <param-value>
  authz:Bearer|.*/api/*
  authz:Basic|.*/api/*
</param-value>
</init-param>
```
1. when the excludePatterns are being checked, if excludePattern startsWith authz then get first indexOf | and use this to check the Authorization Scheme for the excludePattern is Valid
1. for backward compatibility, if no pattern startsWith authz then each excludePattern would continue the  FilterChain
 ---
I'll start working on the excludePatterns enhancement if you agree

pedroneil commented:
@hazendaz I have removed the batch file and use that as my own local utility
I have updated the waffle-jna pom and waffle-tests to use the same mockito-core.version 2.19.0 and all tests still pass.

hazendaz commented:
@pedroneil Perfect!  This looks great!  I'm good with you working on exclude patterns next.  I can get 1.9.1 released this weekend unless you want to wait until you have that work done as well.  I'm going to swatch this on the merge given number of commits.  Thanks.

pedroneil commented:
I'd rather not wait for for 1.9.1. I'd rather be running with a build from Waffle than a snapshot. 
Thanks for the squash.
I'll create a new branch and do the next set of work with fewer commits.

hazendaz commented:
Ok – sounds good.

From: Peter Neil <notifications@github.com>
Sent: Saturday, June 30, 2018 4:03 PM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; State change <state_change@noreply.github.com>
Subject: Re: [Waffle/waffle] Exclude OPTIONS for CORS preflight and Bearer Authorization (#631)


I'd rather not wait for for 1.9.1. I'd rather be running with a build from Waffle than a snapshot.
Thanks for the squash.
I'll create a new branch and do the next set of work with fewer commits.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/631#issuecomment-401563128>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AA7ho1bx5NzXKb6waEXPQdIXB9T1PYPOks5uB9lcgaJpZM4U5S1H>.
